### PR TITLE
Auto-push bundle manifests changes to PR branch if needed

### PR DESCRIPTION
**Pull request by rm3l**
## Description
This is to ensure bundle manifests are kept updated when submitting a PR. 

## Which issue(s) does this PR fix or relate to
This is a follow-up PR to https://github.com/janus-idp/operator/pull/187#issuecomment-1938816003

## PR acceptance criteria

- [ ] Tests
- [x] Documentation

## How to test changes / Special notes to the reviewer
- ~https://github.com/janus-idp/operator/actions/runs/7878373268?pr=195 is an example of run with outdated bundle manifests~
- ~After applying the resulting patch (attached as a job artifact) or after regenerating the bundle with `make bundle`, the job passes successfully, e.g.: https://github.com/janus-idp/operator/actions/runs/7878650096/job/21497344358?pr=195~

Because the workflow is triggered on `pull_request_target` events, this PR will need to be merged into the base branch to see it in action.

### CODE REVIEW

File: .github/workflows/pr.yaml
Comment (gazarenkov) [inline .github/workflows/pr.yaml:57]: To me it looks like adding complexity for pretty much zero outcome, sorry :) 
This file bundle/manifests files are autogenerated by make bundle, so this can NOT "affect the resulting manifests" and so the message looks confusing.

File: bundle/manifests/backstage-default-config_v1_configmap.yaml
Comment (gazarenkov) [inline bundle/manifests/backstage-default-config_v1_configmap.yaml:62]: For my understanding: Why reviewer should spend time reviewing autogenerated files?

**Review by gazarenkov — CHANGES_REQUESTED** [review]

File: .github/workflows/pr.yaml
Comment (rm3l) [reply .github/workflows/pr.yaml:57]: > To me it looks like adding complexity for pretty much zero outcome, sorry :)

No problem - always open to discussions :)
So the whole value here is to alleviate the load on reviewers, using some automation to streamline the review process as much as we can. This could be part of our PR template checklist, but it is always better if it can be enforced automatically, which is the point here.

> I mean I do not mind of calling extra 'make bundle' if we really need

Sure, we can think of calling `make bundle` because we know that, but if someone else contributes a PR, we would need to think about checking the bundle manifests. Even us, didn't we forget sometimes to regenerate the bundle in most of the PRs where it could have been needed? 
Same point: if you update the go code using the kubebuilder marker comments and don't regenerate the CRDs manifests (I think it already happened recently when updating the RBAC permissions), this would be caught by this validation workflow.

> sending the message about autogenerated by 'make bundle' file looks confusing for me , this can NOT "affect the resulting manifests" for github workflow.

Not sure I understand this statement. You mean the message displayed by this step, like in [this run](https://github.com/janus-idp/operator/actions/runs/7878373268/job/21496491114?pr=195#step:5:49)?
```
Files changed in bundle generation.
Please make sure to regenerate the bundle with 'make bundle' and push the changes.
Make sure you unset any related env vars like VERSION or IMAGE_TAG_BASE or IMG before running this command, as they may affect the resulting manifests.
You might also need to update the CSV in '.rhdh/bundle/manifests/rhdh-operator.csv.yaml' file accordingly.
```
This is a hint to the PR author if they want to call `make bundle` by themselves locally, to pay attention to those variables. By "affect the resulting manifests", I mean that if you had exported an env var called `VERSION` or `IMAGE_TAG_BASE`, the generated CSV will likely contain those values. Please try it by yourself, by calling `export IMAGE_TAG_BASE=my.registry.example.com/janus-idp/operator; make bundle`, and you'll see the result.
But I'm open to a different wording suggestion ;)

The CI is the source of truth, so the diff seen by the CI should be the expected outcome. Initially, I was thinking of making the GitHub job just commit the diff, so that the PR author does not even need to think about it. But for now, erroring out in case of a diff will help educate contributors to think about the resulting manifests (because the CSV in `.rhdh` might also need to be changed). That's why I'm attaching the diff to review and apply it as needed.

File: bundle/manifests/backstage-default-config_v1_configmap.yaml
Comment (rm3l) [reply bundle/manifests/backstage-default-config_v1_configmap.yaml]: To me, even if it is generated, this is also part of the code; just like the manifests generated via the [kubebuilder marker comments](https://book.kubebuilder.io/reference/generating-crd). It also seems quite common in the operators world, no? ([opentelemetry-operator](https://github.com/open-telemetry/opentelemetry-operator/tree/main/bundle), [observability-operator](https://github.com/rhobs/observability-operator/tree/main/bundle), [jaeger-operator](https://github.com/jaegertracing/jaeger-operator/tree/main/bundle), [grafana-operator](https://github.com/grafana/grafana-operator/tree/master/bundle), ...)
It also makes it easier to check them with code-scanning tools like SonarCloud, Snyk.

File: .github/workflows/pr.yaml
Comment (jianrongzhang89) [reply .github/workflows/pr.yaml:57]: Enforcing the validation sounds a good idea to me and it is commonly used by others too.

Comment (gazarenkov) [reply .github/workflows/pr.yaml:57]: What really concerning me is that Developer (in most of the cases triggered by this proposal) may not even have an idea that he/she changed something related.
Like I mentioned: for development I use variable like IMG to test it with my repo/image not touching prod one. It is a valid case, right?
So, let's imagine I have no idea that I did something "affecting the resulting manifests" even not changing source code and this confusing me a lot. It is happened due to particular internal implementation details and I do not think everyone should be aware of it.
TBH I thought you will make GH action provide some automation to change the things silently, I even did not consider any messaging back.

**Review by gazarenkov — CHANGES_REQUESTED** [review]

File: docs/developer.md
Comment (jianrongzhang89) [inline docs/developer.md:100]:
```suggestion
make manifests bundle
```

**Review by jianrongzhang89 — CHANGES_REQUESTED** [review]

File: .github/workflows/pr.yaml
Comment (jianrongzhang89) [reply .github/workflows/pr.yaml:57]: @gazarenkov Developers should be aware that local builds can generate local-specific manifests, and before they submit a PR, they should re-generate the manifests/bundle without local specific information. This is what I have been doing and it has been working well for me:
unset IMG && unset IMAGE_TAG_BASE && unset VERSION && make manifests bundle

Comment (rm3l) [reply .github/workflows/pr.yaml:57]: > What really concerning me is that Developer (in most of the cases triggered by this proposal) may not even have an idea that he/she changed something related. Like I mentioned: for development I use variable like IMG to test it with my repo/image not touching prod one. It is a valid case, right? So, let's imagine I have no idea that I did something "affecting the resulting manifests" even not changing source code and this confusing me a lot. It is happened due to particular internal implementation details and I do not think everyone should be aware of it. 

If you did not change anything affecting the resulting manifests, then the job would just pass with no action required from you. Otherwise, you would learn that there is something needed. ;)

> TBH I thought you will make GH action provide some automation to change the things silently, I even did not consider any messaging back.

I was about to do it that way, but thought about the downstream CSV under `.rhdh`, which we may need to update as well. So IMO, we'd better enforce those with clear messaging, so that people know.
It is somehow like the lint checks to me. Sometimes, linters might be able to auto-fix some things, but in CI we tend to prefer erroring out, so that developers know for next time.

For reference, I noticed quite a few operators out there have similar validation checks (generating, then checking the git diff). See:
- https://github.com/rhobs/observability-operator/blob/main/.github/workflows/pr-checks.yaml#L60-L62
- https://github.com/os-observability/tempo-operator/blob/main/Makefile#L397-L407
- https://github.com/jaegertracing/jaeger-operator/blob/main/Makefile#L121-L127
- https://github.com/grafana/grafana-operator/blob/master/.github/workflows/pr-validation.yaml#L166-L176

Comment (gazarenkov) [reply .github/workflows/pr.yaml:57]: >Developers should be aware that local builds can generate local-specific manifests,

Even knowing about bundle generation specific, Developers must not keep everything in the head, keeping all the concerns unseparated is not the best way to design.
Failing workflow for fully workable code just because developer did not perform 'make bundle' locally is too confusing especially taking into account that OLM is not the only way of deploying upstream operator (even to Opeshift).

>but thought about the downstream CSV under .rhdh, which we may need to update as well.

Not sure I understand, I thought .rhdh uses manually created manifest, i.e. 'make bundle' makes no diff for it. Did I miss something?

Comment (rm3l) [reply .github/workflows/pr.yaml:57]: > Even knowing about bundle generation specific, Developers must not keep everything in the head, keeping all the concerns unseparated is not the best way to design.
Failing workflow for fully workable code just because developer did not perform 'make bundle' locally is too confusing especially taking into account that OLM is not the only way of deploying upstream operator (even to Openshift).

So that's why enforcing this in CI helps the developer IMO.
Developers are free to iterate no matter how they want locally, but we need to enforce checks after a `git push`, and this includes making sure that it meets the requirements to use the operator in production, e.g.: bundle manifests are consistent with the code changes, potential scanning tools against those manifests, ...
Sure, OLM is not the only way, but it is part of what we ship at the end of the day. So we need to enforce that it can be used consistently there as well.

> Not sure I understand, I thought .rhdh uses manually created manifest, i.e. 'make bundle' makes no diff for it. Did I miss something?
Yes, it is manually updated, but one way to remind us to consider doing so is by highlighting potential changes to the upstream CSV.

Comment (gazarenkov) [reply .github/workflows/pr.yaml:57]: I am OK for warning, letting developer know that it seems the locally generated code does not look the same as it is generated by CI (with possible reasons as mentioned). 

But I disagree of failing workflow for that. Unlike lint (you brought as an analogy):
- developer may not directly deal and care about this aspect (see my example above)
- this has no impact on the end result since CI's "make bundle" will produce exactly the same for synced and un-synced locally generated files

File: bundle/manifests/backstage-default-config_v1_configmap.yaml
Comment (jianrongzhang89) [reply bundle/manifests/backstage-default-config_v1_configmap.yaml]: Yes, generated code should also be reviewed. The fact that the code is generated does not mean it works as intended.

**Review by jianrongzhang89 — APPROVED** [review]
Sounds good to me - a good practice for consistency and code quality.

File: .github/workflows/pr.yaml
Comment (rm3l) [reply .github/workflows/pr.yaml:57]: > I am OK for warning, letting developer know that it seems the locally generated code does not look the same as it is generated by CI (with possible reasons as mentioned).
> 
> But I disagree of failing workflow for that. Unlike lint (you brought as an analogy):
> 
> * developer may not directly deal and care about this aspect (see my example above)
> * this has no impact on the end result since CI's "make bundle" will produce exactly the same for synced and un-synced locally generated files

I am not sure a warning in the validate job would be noticed, unfortunately. I think it is preferable to error out by default (since wrong bundle manifests will affect what we ship in the end).
And with the Prow bot, it is possible to override a required check manually if needed (using the `/override` comment, and providing a justification).
Again, if a developer makes changes that don't affect the manifests, nothing will change for them. The validate job will just pass. But if a change is made that changes the manifests, it would be better to know and validate it IMO.

Comment (gazarenkov) [reply .github/workflows/pr.yaml:57]: I understand your point but it does not look friendly and make no sense for developers who will want to try it w/o OLM, which we suggest as a quick start in fact.
For ones who care about it, warning would probably give some hint (TBH, I'd see some documentation emphasising this case) but error is too much b/c it does not impact CI.

File: bundle/manifests/backstage-default-config_v1_configmap.yaml
Comment (gazarenkov) [reply bundle/manifests/backstage-default-config_v1_configmap.yaml]: You can review whatever you want but, honestly, how often you review for example zz_generated.deepcopy.go file?

If you tend not trust the tool the best thing is go and check it once (maybe trying theirs tests etc), otherwise it become nightmare.
Thats how separation of concern works I think.

File: .github/workflows/pr.yaml
Comment (rm3l) [reply .github/workflows/pr.yaml:57]: Ok, let me check if we can display a warning instead.. We'll see how it goes with a warning.

Comment (rm3l) [reply .github/workflows/pr.yaml:57]: @gazarenkov This should be a warning now ([`011654e` (#195)](https://github.com/janus-idp/operator/pull/195/commits/011654e3399aa3d8e7b9300d5f076e1ce8731a2b)), which won't change the outcome of the "PR Validate" job. Let's see how it goes. We can always change it back to an error in the future if it does not help that much.
To make the warning more noticeable to PR authors and reviewers, I was planning to automatically add a comment to the PR, but it does not work currently on fork PRs due to permission restrictions. This can be fixed in a subsequent issue/PR. For now, PR authors/reviewers should think about checking the validation job summary in GH for any warnings.
See https://github.com/janus-idp/operator/actions/runs/7959583931?pr=195 for an example and approve if it looks fine to you.

image

Comment (nickboldt) [reply .github/workflows/pr.yaml:57]: I would agree that **warning are ignored until they're escalated to errors**. Consider the warnings about deprecated versions of nodejs used in GH actions. Until the action fails, no one is going to look at the warnings and update all their action dependencies to newer versions depending on non-EOL nodejs versions. 

image

-- https://github.com/janus-idp/operator/actions/runs/7960349847

**Review by nickboldt — CHANGES_REQUESTED** [review]
if there's a problem with validation, FAIL the build. Warnings are tantamount to :thisisfine: and no one will react.

File: .github/workflows/pr.yaml
Comment (gazarenkov) [reply .github/workflows/pr.yaml:57]: @nickboldt 
 I may miss something but do not think impact of this PR's warning and warnings about deprecated node.js version are comparable.
 
Let's consider something closer:  https://github.com/janus-idp/operator/issues/192
Would you consider it as "warning are ignored" and "escalate it to error" ?
If so, what concrete action items would you propose?

File: .github/workflows/pr.yaml
Comment (rm3l) [reply .github/workflows/pr.yaml:83]: > it looks like the major part of pr.yaml is dedicated to the case which even does not impact in the end result (it is only about the file which developer generate or does not generate locally)?

Not sure I completely understand your question. You mean a developer making any changes that potentially affect the manifests (even though they are not aware of that in the beginning)?
In that case, I think it should be the responsibility of the CI to make users aware of that fact so that they can fix up their PR. Doing so locally only might be easy to forget or ignore (on purpose or not).

To clarify, we want to prevent the issue where a developer forgets to regenerate the bundle manifests (and potentially update the one in `.rhdh`), causing a mismatch between the code and those manifests.
Regardless of how (error vs warning) we want to make developers aware of such a potential issue, I think it makes sense to enforce this in CI, not only locally.

File: .github/workflows/pr.yaml
Comment (nickboldt) [reply .github/workflows/pr.yaml:57]: my point is that if you want people to notice a problem when running a GH action, it needs to FAIL, not WARN.

Warnings are ignored; failures are addressed.

File: .github/workflows/pr.yaml
Comment (nickboldt) [reply .github/workflows/pr.yaml:83]: >  we want to make developers aware of such a potential issue, I think it makes sense to enforce this in CI

+1. 

The only way to block the PR from being merged is to fail the GH action if a mismatch occurs. A warning will never be seen by the PR reviewer.

File: .github/workflows/pr.yaml
Comment (nickboldt) [inline .github/workflows/pr.yaml:60]: I guess a comment is at least visible to the PR review process, but if action is required to change the PR before merging, then why not FAIL the action and FORCE the developer to fix their PR, instead of just suggesting what needs doing via a comment?

File: .github/workflows/pr.yaml
Comment (nickboldt) [inline .github/workflows/pr.yaml:80]: if you can generate a patch, why not just push a commit with this change?

File: .github/workflows/pr.yaml
Comment (nickboldt) [inline .github/workflows/pr.yaml:58]: see comment above -- `gosec needs a "build" stage so connect it to the lint step which we always do`

File: .github/workflows/pr.yaml
Comment (nickboldt) [inline .github/workflows/pr.yaml:112]:
```suggestion
      - name: build
```

File: .github/workflows/pr.yaml
Comment (rm3l) [reply .github/workflows/pr.yaml:80]: Yeah, initially, I was thinking of making the GitHub job just commit the diff, so that the PR author does not even need to think about it (Gennady's concern, I think).
But then I thought that the CSV in .rhdh might also need to be changed, and I am not sure we can handle that automatically.
Now with hindsight, I think it can indeed make sense to:
- automatically commit any diff with this change, which should sort out @gazarenkov's main concern, I think
- comment on the PR (or update the PR template) that the CSV in .rhdh should also be checked

WDYT? @gazarenkov @jianrongzhang89 @nickboldt

Comment (rm3l) [reply .github/workflows/pr.yaml]: Okay, I've gone ahead and implemented as proposed above:
- bundle manifests changes will be automatically pushed to the PR branch. Due to permission restrictions when using the built-in `GITHUB_TOKEN` token, this had to be done in a separate workflow triggered on `pull_request_target` events (+ same safeguard permissions we are already using to prevent PWN requests from external PRs)
- comment on the PR + update to the PR template that the CSV in `.rhdh` should also be checked

PTAL

/retitle Auto-generate bundle manifests changes to PR branch if needed

File: .github/workflows/pr-bundle-diff-checks.yaml
Comment (nickboldt) [inline .github/workflows/pr-bundle-diff-checks.yaml:1]:
```suggestion
# Copyright 2024 The Janus IDP Authors
```

**Review by nickboldt — APPROVED** [review]
seems reasonable, let's run it and see what explodes :D

File: docs/developer.md
Comment (gazarenkov) [inline docs/developer.md:98]: To cover all the deployment options we offer, I would say something like:
If you are editing the API definitions, make sure you :
- run 'make install' before deploying operator with 'make deploy'
- regenerate the manifests and bundle if you plan to deploy operator with OLM using: ...

**Pull request merged**
